### DISCUSSION TITLE
Question: [Is there a plan/schedule for adding InkCanvas into WinUI desktop?]

### DISCUSSION THREAD

**Discussion by ranbims** [microsoft/microsoft-ui-xaml#7426] [Q&A]
[This document](https://docs.microsoft.com/en-us/windows/apps/windows-app-sdk/migrate-to-windows-app-sdk/what-is-supported) says that `InkCanvas` is not supported in WinUI 3. Just want to know is there a plan for supporting this feature?

**Comment by castorix** [comment]
A workaround is with the **InkPicture** control
But I had to use a Win32 Region to make it appear :

![InkPicture](https://user-images.githubusercontent.com/22345506/154687250-1b77d0e6-241a-43f2-848d-d0c2b8345b9b.gif)

**Reply by DanielGlick** [reply]
Can you give direction on how this is done? We are trying to move from UWP to WinUI 3 and there is currently no way to use inking features in WinUI 3

**Reply by castorix** [reply]
> Can you give direction on how this is done? We are trying to move from UWP to WinUI 3 and there is currently no way to use inking features in WinUI 3

I used ATL, like the test sample with MSRDP ([WinUI3_ActiveX_MSRDP](https://github.com/castorix/WinUI3_ActiveX_MSRDP))
but with "_msinkaut.InkPicture.1_" (3rd parameter for _CreateWindowEx_), which is the ProgID for the InkPicture control

**Reply by DanielGlick** [reply]
Got it. We ended up using another workaround by drawing lines on a grid based on pen/touch input

**Comment by kmgallahan** [comment]
WinUI roadmap: https://github.com/microsoft/microsoft-ui-xaml/blob/main/docs/roadmap.md

InkCanvas on the product board: https://portal.productboard.com/winappsdk/1-windows-app-sdk/c/31-inking-controls?utm_medium=social&utm_source=portal_share

**Comment by ranbims** [comment]
> A workaround is with the **InkPicture** control But I had to use a Win32 Region to make it appear :
> 
> ![InkPicture](https://user-images.githubusercontent.com/22345506/154687250-1b77d0e6-241a-43f2-848d-d0c2b8345b9b.gif)

Thank you @castorix. I am wondering if I can use some workaround to show the **UWP** or **WPF** InkCanvas here.

**Comment by ranbims** [comment]
> WinUI roadmap: https://github.com/microsoft/microsoft-ui-xaml/blob/main/docs/roadmap.md
> 
> InkCanvas on the product board: https://portal.productboard.com/winappsdk/1-windows-app-sdk/c/31-inking-controls?utm_medium=social&utm_source=portal_share

Thanks, @kmgallahan . Glad to see the InkCanvas is planned for future updates. I am curious if there is a specific or an estimated date when the feature will be released. We are planning a migration for a current UWP application. And it might be a ship blocker if we can't enable the Ink related features.

**Reply by soroshsabz** [reply]
ITNOA

I think it is good to add feature proposal for this

https://github.com/microsoft/microsoft-ui-xaml/blob/main/docs/feature_proposal_process.md

**Comment by StephenLPeters** [comment]
@ryandemopoulos FYI

**Comment by DilanBoskan** [comment]
Please add this. My whole team solely requires the InkCanvas control to move to the Windows App SDK.

**Comment by nikolayvpavlov** [comment]
InkCanvas is a must.  Our company are to develop an app for onsite inspections, and taking notes is an important feature.  

InkPicture is not suitable for text recognition.  Perhaps it can still be used if I connect to Stroke event and record all strokes myself, but that is not nice.

I managed to get it running by letting the user draw on a canvas and I convert the lines to strokes, and then use InkAnalysis to convert them to text.  However, performance is slow, and stokes look ugly on the screen, even though I simulate the pen pressure effect.

**Reply by castorix** [reply]
> InkPicture is not suitable for text recognition. 

MS samples seem to work correctly, like :

![InkPicture_reco](https://github.com/microsoft/microsoft-ui-xaml/assets/22345506/ec530be1-df02-4ce9-aeca-43340d8a4a19)

or

![InkPicture_reco2](https://github.com/microsoft/microsoft-ui-xaml/assets/22345506/c491fe00-c49b-4730-931b-00f708433ba0)

**Comment by jdluzen** [comment]
Are there any updates on this? I am currently awaiting Ink support for InfiniteCanvas in WinUI3. https://github.com/CommunityToolkit/WindowsCommunityToolkit/issues/4105#issuecomment-878431252

**Comment by SID9-HoneyBee** [comment]
Using InkCanvas
There is a clear difference between developing with WPF and developing with UWP. (The lines are much smoother and there is no distortion.)

Any idea when will InkCanvas be introduced in WinUI3?

Looking forward to it very much.

**Reply by TheAjaykrishnanR** [reply]
https://github.com/microsoft/microsoft-ui-xaml/assets/110657810/e7445f6e-d7bf-4100-b351-5ede6faf8ac3

Drawing lines and then drying them using a Path (Microsoft.UI.Xaml.Shapes) with PolyBezier Geometry can help.

**Reply by SID9-HoneyBee** [reply]
@TheAjaykrishnanR I think it's a good approach. Is it possible to adjust the tension for Bezier curve processing?

**Reply by TheAjaykrishnanR** [reply]
@SID9-HoneyBee I am not sure about a "tension", but maybe you can achieve that affect specifying the number of points to build the geometry, reducing those points might just do that, this is the example I used :
https://learn.microsoft.com/en-us/dotnet/api/system.windows.media.polybeziersegment?view=windowsdesktop-8.0

**Reply by DilanBoskan** [reply]
This just isn't as performant as the native InkCanvas though. The native InkCanvas uses a BackgroundThread to capture the inputs and as far as I am concerned a more sophisticated rendering technique.
In addition, the features of the native InkCanvas, of which I make use of thoroughly, are also not included.

**Reply by TheAjaykrishnanR** [reply]
@DilanBoskan true ! it has a very high latency, and if you observe correctly the endpoints dont even have rounded caps, i had to add ellipses to make it look rounded. had to revert back to uwp

**Reply by SID9-HoneyBee** [reply]
@DilanBoskan @TheAjaykrishnanR 

I agree with you.
The InkCanvas of WPF and UWP are clearly different. UWP is much smoother for the same lines.
What do you think?
I would like to achieve the same result in WPF. I'd also like to include the ability to fine-tune the FitToCurve property, but it's really not that easy.

**Comment by Neme12** [comment]
Is it not possible to use XAML Islands to embed the UWP InkCanvas control in a WinUI 3 desktop app?

**Reply by TheAjaykrishnanR** [reply]
I dont think so, XAML Islands are for WPF, Win32 apps etc, not for WinUI
https://learn.microsoft.com/en-us/windows/apps/desktop/modernize/xaml-islands

**Reply by DarranRowe** [reply]
Not without the support of WinUI 3 itself.
The inking controls would need to  be wrapped in a Windows HWND, and this would have to be placed inside the WinUI 3 content bounds. WinUI 3 would have to handle this in the same way as controls like web view, as described in the [WAS Visual Layer](https://learn.microsoft.com/en-us/windows/apps/windows-app-sdk/composition#differences-from-uwp) documentation.
While it is is possible to implement this, I believe that it would end up with a better control if it was implemented natively as a WinUI 3 control.

**Reply by TheAjaykrishnanR** [reply]
@DarranRowe So when you use something like a swapchain onto which you render the inkcanvas, what exactly is handling the input processing ? I mean if it is the composite surface that is passing on the pointer inputs to the inkcanvas then would it be any better ?

**Reply by DarranRowe** [reply]
Well, I thought I implied it already. But

>Not without the support of WinUI 3 itself.

and 

>WinUI 3 would have to handle this in the same way as controls like web view,

would mean that it would be the WinUI 3 window handling the pointer inputs and dispatching them correctly. To be explicit, I am not saying that someone could just add the UWP ink control in a HWND to any WinUI 3 window now. I am saying that, I believe that to get the UWP version inside an Xaml Island working properly with WinUI 3, then WinUI 3 would have to take on the burden of handling things in the same way as Web View 2.

**Reply by TheAjaykrishnanR** [reply]
> it would be the WinUI 3 window handling the pointer inputs and dispatching them correctly.

But I dont think thats how InkCanvas works, i.e. it is not utilizing the window's pointer event handling thread to capture its inputs. Rather it is using its own low latency background thread to capture it thereby making it faster and snappier.

**Reply by DarranRowe** [reply]
Yes, but even then it would have to get the input from somewhere.
The normal way of doing that would be through the thread's message handling. So Xaml Islands would use [IDesktopWindowXamlSourceNative2:PreTranslateMessage](https://learn.microsoft.com/en-us/windows/win32/api/windows.ui.xaml.hosting.desktopwindowxamlsource/nf-windows-ui-xaml-hosting-desktopwindowxamlsource-idesktopwindowxamlsourcenative2-pretranslatemessage) to notify the Xaml Islands window of the various events happening on the thread. This would have to be in the message pump, and for a WinUI 3 application, this is started as part of Application.Start. So either way, WinUI 3 would need to handle the Xaml Island window properly.
If it is using some other means of monitoring the input, then there is no reason why that wouldn't work anyway making that whole point moot. Remember, because of the interoperation going on, the CoreWindow has to exist on the thread.

--Edit--
The reason that I am pretty confident in this is due to the documentation for Xaml Islands and the requirements for C++ applications. The only real requirement for InkCanvas and InkToolbar is a fully working DesktopWindowXamlSource. The only documented limitation to ink input is that HandwritingView isn't supported.

**Reply by Neme12** [reply]
> I dont think so, XAML Islands are for WPF, Win32 apps etc, not for WinUI https://learn.microsoft.com/en-us/windows/apps/desktop/modernize/xaml-islands

But WinUI 3 apps *are* Win32 apps. It should be possible to do. It just won't be as easy as adding a pre-built XAML wrapper control.

**Reply by TheAjaykrishnanR** [reply]
> > I dont think so, XAML Islands are for WPF, Win32 apps etc, not for WinUI https://learn.microsoft.com/en-us/windows/apps/desktop/modernize/xaml-islands
> 
> But WinUI 3 apps *are* Win32 apps. It should be possible to do. It just won't be as easy as adding a pre-built XAML wrapper control.

So apparently, you can, but i have no expertise in C++/WinRT

**Reply by castorix** [reply]
> Is it not possible to use XAML Islands to embed the UWP InkCanvas control in a WinUI 3 desktop app?

I did a test and it worked, but I had to use a Win32 process to create XAML Islands with InkCanvas (+ InkToolbar) in a WinUI 3 unpackaged app (otherwise I get "80073D54 - The process has no package identity" if I call  _"WindowsXamlManager xamlManager = WindowsXamlManager.InitializeForCurrentThread();"_ or create an instance of _Windows.UI.Xaml.Hosting.DesktopWindowXamlSource_), then I used SetParent to embed it :

![WinUI3_InkCanvas](https://github.com/user-attachments/assets/2db9a9a4-ccde-4420-b2b8-8ae97729b66a)

**Reply by castorix** [reply]
I just uploaded a test sample https://github.com/castorix/Winui3_XAMLIslands_InkCanvas with XAML Islands, mainly to check if basic functionalities could work, but it is better of course to wait WinUI 3 InkCanvas in stable releases...

**Comment by nikolayvpavlov** [comment]
The InkCanvas is there in the experimental releases (1.7 and 1.8), but not included in the stable releases.  I tried it and it works quite well, although the API is slightly different than UWP and there are a few quirks that may be confusing without proper documentation.

**Reply by castorix** [reply]
> The InkCanvas is there in the experimental releases (1.7 and 1.8), but not included in the stable releases. I tried it and it works quite well, although the API is slightly different than UWP and there are a few quirks that may be confusing without proper documentation.

Yes, I tried it too with 1.8.250507002-experimental and it works fine, like UWP InkCanvas (InkPresenter got from QueueInkPresenterWorkItem)
But InkToolbar is not yet in Microsoft.WinUI.dll

**Comment by Skittles2519** [comment]
I'm still unable to publish an update to the store with the InkCanvas since its still in experimental. Can we please get some motion on this?

**Reply by DilanBoskan** [reply]
Yes.. please, we are waiting for years to migrate off of UWP!

**Comment by smourier** [comment]
For those like me waiting for this for sooo long 😭, here's a full C# "portable" implementation sample https://github.com/smourier/WinUI3Ink

It works in WinAppSDK release channels (2.0+), and it has no toolbar (but you can build your own)

**Reply by TheAjaykrishnanR** [reply]
very nice ! works flawlessly

**Reply by Skittles2519** [reply]
Wow thanks for sharing this! I was able to swap the control out fairly easily. I was not smart enough to get the other work arounds to work lol

**Reply by gautambjain** [reply]
Thanks a lot. How can I easily save the drawing to a file?

**Reply by smourier** [reply]
> Thanks a lot. How can I easily save the drawing to a file?

I've enhanced the sample to support this.
